WordPress/openverse
# Use a hack to compile the frontend container with workspaces

## Description




This PR updates the `Dockerfile` and the `build-frontend` job to
- use the frontend locally as a part of the workspace
- build the Docker image as a non-workspace JS package

To do this, the PR employs a hack whereby a monorepo like structure is set up inside the Docker image itself by outer using files from the monorepo like `.npmrc`, `pnpm-lock.yaml` and `.pnpmfile.cjs`. Also a fake `pnpm-workspaces.yaml` file is generated that only contains on sub package, the frontend itself.

I have verified that an image generated in this way is very similar in size to the non-workspaced image. Also the root lockfile has been pinned in a way that very much prefers the package versions specified in the previous frontend-only lockfile (notably Nuxt and Vue packages) so that accidental updates are not introduced.

## Testing Instructions

Download the `frontend.zip` artifact from the CI + CD summary page. Then extract and load it into Docker as follows.

```console
$ unzip frontend.zip
$ docker load --input frontend.tar
$ docker image ls
```

You can start the Nuxt server from the downloaded image. See the frontend running properly at http://localhost:50243.

```console
$ docker run --rm -p 50243:8443 openverse-frontend:latest
```


## Checklist



      `Update index.md`).
      a parent feature branch.
      errors.

[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin



<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
